### PR TITLE
Registry per Cluster prototype

### PR DESCRIPTION
**Pull request by nachocano**
## Proposed Changes

- Adding EventType CRD with cluster scope.
- Configuring few ingress policies in Broker.
- RBAC permissions for ingress to retrieve eventtypes (not entirely working at the cluster level).

### CODE REVIEW

**Comment by nachocano** [comment]
@Harwayne
@grantr

As FYI, an initial version of the registry I've been working on..
Will take a look at how events are being sent by our current sources.

**Comment by nachocano** [comment]
FYI @akashrv

File: pkg/broker/ingress.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+	}
+}
+
+func (policy *AllowAnyPolicy) AllowMessage(namespace string, message *provisioners.Message) bool {
+	return true
+}
+
+func (policy *AllowRegisteredPolicy) AllowMessage(namespace string, message *provisioners.Message) bool {
+	eventType := &eventingv1alpha1.EventType{}
+	name := message.Headers["Ce-Eventtype"]
```
Comment (grantr) [inline pkg/broker/ingress.go]: This should probably be a constant.

According to https://github.com/cloudevents/spec/blob/master/http-transport-binding.md, the header name in the HTTP transport is `ce-type`. Is this transformed to `Ce-Eventtype` somewhere else, perhaps in the CloudEvents SDK?

**Comment by grantr** [comment]
This seems fine to me, but I'll let @harwayne do the final review. Thanks @nachocano!

File: pkg/broker/ingress.go (hunk 1 above)
Comment (nachocano) [reply pkg/broker/ingress.go]: Yes, I agree. Should be some constant probably.

I asked Scott, he said that ce-type is spec version 0.2. For 0.1 is called Ce-Eventtype.

Comment (nachocano) [reply pkg/broker/ingress.go]: Actually, Scott suggested we should use the SDK to decode the message. 
I'll do those changes... We need to do it as well for the filtering in triggers...

**Comment by nachocano** [comment]
@Harwayne
@grantr 

Will add changes to use the SDK to decode the event... so we don't have to care about those constants... Will add that change in the trigger filter prototype we have as well..

**Comment by nachocano** [comment]
I take that back. Seems that is a bigger change than I thought. We have the Message struct everywhere. Was wondering whether we should cloudevent.Event. 
Will talk to you guys in person, will leave it as is for now, and try to understand better the sources.

**Comment by nachocano** [comment]
@Harwayne
@grantr 

I was able to send events from four different sources (k8s, cronjob, gcppubsub, github). Attached are the yamls of the sources. At the end of each file you can see received events in the dumper.
For github I was listening to pull_request and push. I created a PR, and then did a push in the same PR branch, but the event triggered was still a pull_request, but with a different action (synchronize). You can see it in the log.
All are Cloud Events v0.1. I guess we are converting them to that spec version somewhere. If we will keep on doing that, then having that constant for Ce-Eventtype and Ce-Source might be good enough. Otherwise I think we should try to include Scott's cloudevent SDK to parse the message.

[eventsFromSources.zip](https://github.com/nachocano/eventing/files/2917512/eventsFromSources.zip)

File: cmd/broker/ingress/main.go [hunk 2]
```diff
@@ -156,3 +188,23 @@ func (r *runnableServer) Start(<-chan struct{}) error {
 	r.logger.Info("Ingress Listening...", zap.String("Address", r.s.Addr))
 	return r.s.ListenAndServe()
 }
+
+// TODO this should be removed once we update the interfaces and start using cloudevents.Event instead of Message.
+func cloudEventFrom(m *provisioners.Message) cloudevents.Event {
```
Comment (n3wscott) [inline cmd/broker/ingress/main.go:195]: oh man you should use the cloudevents lib: https://github.com/cloudevents/sdk-go

Comment (nachocano) [reply cmd/broker/ingress/main.go:195]: Yes, I put that in the TODO. 
BTW how do I convert the provisioners.Message to a cloudevents.Event if I don't inspect the headers? 
Adam was working on changing the API to not receive a Message as an arg, but a cloudevent... I was waiting for that change

**Comment by nachocano** [comment]
Won't pursue this one. Already have a working version of Registry per namespace, which seems more suitable for our needs in https://github.com/nachocano/eventing/pull/7.
Closing this PR for now

**Pull request closed without merging**